File: artifacts/ht/translate/src/lsearch.rs
use translate_new::*;
use std::env;
use std::fmt;

pub fn main() -> Result<(), Box<dyn std::error::Error>> {
    let items = [
        Item { key: Some("foo".to_string()), value: 10 },
        Item { key: Some("bar".to_string()), value: 42 },
        Item { key: Some("bazz".to_string()), value: 36 },
        Item { key: Some("buzz".to_string()), value: 7 },
        Item { key: Some("bob".to_string()), value: 11 },
        Item { key: Some("jane".to_string()), value: 100 },
        Item { key: Some("x".to_string()), value: 200 },
    ];
    let num_items = items.len();
    let found = linear_search(&items, Some("bob"));
    
    match found {
        Some(item) => println!("linear_search: value of 'bob' is {}", item.value),
        None => return Err(Box::new(fmt::Error)),
    }
    
    Ok(())
}

File: artifacts/ht/refactor/lsearch.c
#include <lsearch.c>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

typedef struct 
{
  char *key;
  int value;
} item;
item *linear_search(item *items, size_t size, const char *key)
{
  for (size_t i = 0; i < size; i += 1)
  {
    if (strcmp(items[i].key, key) == 0)
    {
      return &items[i];
    }
  }

  return 0;
}

int main(void)
{
  item items[] = {{"foo", 10}, {"bar", 42}, {"bazz", 36}, {"buzz", 7}, {"bob", 11}, {"jane", 100}, {"x", 200}};
  size_t num_items = (sizeof(items)) / (sizeof(item));
  item *found = linear_search(items, num_items, "bob");
  if (found == 0)
  {
    return 1;
  }
  printf("linear_search: value of 'bob' is %d\n", found->value);
  return 0;
}


typedef struct 
{
  char *key;
  int value;
} item;
item *linear_search(item *items, size_t size, const char *key)
{
  for (size_t i = 0; i < size; i += 1)
  {
    if (strcmp(items[i].key, key) == 0)
    {
      return &items[i];
    }
  }

  return 0;
}

int main(void)
{
  item items[] = {{"foo", 10}, {"bar", 42}, {"bazz", 36}, {"buzz", 7}, {"bob", 11}, {"jane", 100}, {"x", 200}};
  size_t num_items = (sizeof(items)) / (sizeof(item));
  item *found = linear_search(items, num_items, "bob");
  if (found == 0)
  {
    return 1;
  }
  printf("linear_search: value of 'bob' is %d\n", found->value);
  return 0;
}

